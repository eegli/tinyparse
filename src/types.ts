export type FlagOption = {
  required?: boolean;
  description?: string;
  shortFlag?: `-${string}`;
};

export type FilePathArg = {
  longFlag: `--${string}`;
  description?: string;
};

export type ParsingOptions<
  K = string,
  V = K extends Record<string, unknown> ? keyof K : K extends string ? K : never
> = {
  filePathArg?: FilePathArg;
  options?: {
    [K in Extract<V, string>]?: FlagOption;
  };
};

export type InternalOptions = (FlagOption & { name: string })[];

export type SimpleRecord<T extends string = string> = Record<T, ObjectValues>;

export type ObjectValues = string | number | boolean;
