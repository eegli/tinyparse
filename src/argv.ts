import { Parser } from './parser';
import { PositionalArgs, SimpleRecord, WithPositionalArgs } from './types';
import { parseJSONFile } from './utils';

type TransformArgs = {
  aliases: Map<string, string>;
  filePathFlag?: string;
};

export class ArgvParser<T extends SimpleRecord> extends Parser<T> {
  public build(input: T, positionals: string[]): WithPositionalArgs<T> {
    return {
      ...input,
      _: positionals,
    };
  }
  public transform(
    argv: string[],
    { aliases, filePathFlag }: TransformArgs
  ): [Partial<T>, PositionalArgs] {
    const flagMap = new Map<string, unknown>();

    const positionals: string[] = [];
    let isPositional = true;

    for (let i = 0; i < argv.length; i++) {
      let curr = argv[i];

      if (!curr.startsWith('-') && isPositional) {
        positionals.push(curr);
        continue;
      }

      isPositional = false;

      // Convert short flag to long flag
      if (aliases.has(curr)) {
        curr = '--' + aliases.get(curr);
      }

      // Ignore non-flags
      if (!curr.startsWith('--')) {
        continue;
      }

      let flag = curr.slice(2);
      if (aliases.has(flag)) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        flag = aliases.get(flag)!;
      }

      const flagVal = argv[i + 1];

      // Parse a file
      if (filePathFlag?.slice(2) === flag) {
        parseJSONFile(flagVal).forEach(([key, content]) =>
          flagMap.set(key, content)
        );
      }
      // Assume boolean flag
      else if (!flagVal || flagVal.startsWith('-')) {
        flagMap.set(flag, true);
        // Assume number
        // Assume string
      } else {
        flagMap.set(flag, flagVal);
      }
    }

    return [Object.fromEntries(flagMap) as Partial<T>, positionals];
  }
}
