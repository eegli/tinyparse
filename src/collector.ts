import { WithPositionalArgs } from './types';

export class Collector<T> {
  constructor(private readonly defaults: T) {}
  public collectWithPositionals(positionals: string[]): WithPositionalArgs<T> {
    return {
      ...this.collect(),
      _: positionals,
    };
  }

  public collect(): T {
    return this.defaults;
  }
}
