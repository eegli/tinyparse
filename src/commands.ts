import { CommonConfig } from './config';
import { Parser } from './parser';
import {
  AnyGlobal,
  CommandArgPattern,
  DefaultHandler,
  Downcast,
  FlagOptionValue,
  FlagOptions,
  FlagValueRecord,
  MetaOptions,
  Subcommand,
} from './types';

export class CommandBuilder<
  Options extends FlagValueRecord,
  Globals extends AnyGlobal,
> {
  #config: CommonConfig<Options, Globals> = {
    meta: {},
    options: new Map(),
    commands: new Map(),
    globalSetter: () => ({}) as Globals,
    defaultHandler: () => {},
  };

  #takenFlags = new Set<string>();

  #validateCommand = (command: string): void => {
    if (this.#config.commands.has(command)) {
      throw new Error(`Command "${command}" has been declared twice`);
    }

    if (
      command === this.#config.meta.help?.command ||
      command === this.#config.meta.version?.command
    ) {
      throw new Error(
        `Subcommand "${command}" has already been declared as a help or version command`,
      );
    }
  };

  #validateOption = (
    key: string,
    longFlag: string,
    shortFlag?: string,
  ): void => {
    if (this.#config.options.has(key)) {
      throw new Error(`Option "${key}" has been declared twice`);
    }

    if (this.#takenFlags.has(longFlag)) {
      throw new Error(`Long flag "${longFlag}" has been declared twice`);
    }
    if (!shortFlag) return;

    if (this.#takenFlags.has(shortFlag)) {
      throw new Error(`Short flag "${shortFlag}" has been declared twice`);
    }
  };

  #validateIdentifiers = (
    command?: string,
    flags?: (string | undefined)[],
  ): void => {
    if (command && this.#config.commands.has(command)) {
      throw new Error(`"${command}" has already been declared as a subcommand`);
    }
    for (const flag of flags || []) {
      if (flag && this.#takenFlags.has(flag)) {
        throw new Error(`"${flag}" has already been declared as a flag`);
      }
    }
  };

  option<K extends string, V extends FlagOptionValue>(
    key: K,
    opts: FlagOptions<V>,
  ) {
    const { longFlag, shortFlag } = opts;
    this.#validateOption(key, longFlag, shortFlag);
    this.#config.options.set(key, opts);
    this.#takenFlags.add(longFlag);
    if (shortFlag) this.#takenFlags.add(shortFlag);

    // TODO: Figure out how to make this typecheck properly
    return this as unknown as CommandBuilder<
      Options & Record<K, Downcast<V>>,
      Globals
    >;
  }

  subcommand<A extends CommandArgPattern>(
    command: string,
    opts: Subcommand<Options, Globals, A>,
  ) {
    this.#validateCommand(command);
    this.#config.commands.set(command, opts);
    return this;
  }

  setGlobals<G extends Globals>(setGlobals: (options: Options) => G) {
    this.#config.globalSetter = setGlobals;
    return this as unknown as CommandBuilder<Options, G>;
  }

  setMeta(meta: MetaOptions) {
    this.#validateIdentifiers(meta.help?.command, [
      meta.help?.longFlag,
      meta.help?.shortFlag,
    ]);
    this.#validateIdentifiers(meta.version?.command, [
      meta.version?.longFlag,
      meta.version?.shortFlag,
    ]);
    for (const flag of [
      meta.help?.longFlag,
      meta.help?.shortFlag,
      meta.version?.longFlag,
      meta.version?.shortFlag,
    ].filter(Boolean) as string[]) {
      this.#takenFlags.add(flag);
    }

    this.#config.meta = meta;
    return this;
  }

  defaultHandler(handler?: DefaultHandler<Options, Globals>) {
    return new Parser<Options, Globals>({
      ...this.#config,
      defaultHandler: handler || this.#config.defaultHandler,
    });
  }
}
