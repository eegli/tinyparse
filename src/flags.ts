import { ValidationError } from './error';
import { FlagDefaultValue, FlagOptions } from './types';
import Utils, { Type } from './utils';

export const collect = (
  inputFlags: Map<string, string | null>,
  flagOptions: Map<string, FlagOptions>,
) => {
  const output = new Map<string, FlagDefaultValue>();
  for (const [key, opts] of flagOptions) {
    const { longFlag, shortFlag, required, defaultValue } = opts;
    // Try to match a long flag - null is a valid flag argument which
    // occurs as a shortcut for boolean flags
    let flagArg = inputFlags.get(longFlag);
    if (flagArg === undefined && shortFlag) {
      // Try to match a short flag
      flagArg = inputFlags.get(shortFlag);
    }
    // The flag is not present but required
    if (flagArg === undefined) {
      if (required) {
        throw new ValidationError(`Missing required option ${longFlag}`);
      }
      output.set(key, defaultValue);
      continue;
    }

    const expectedArgType = Utils.typeof(defaultValue);
    const argumentIsNull = flagArg === null;

    if (expectedArgType === Type.String) {
      if (!argumentIsNull) {
        output.set(key, flagArg as string);
        continue;
      }
      throw new ValidationError(`${longFlag} expects an argument`);
    }
    if (expectedArgType === Type.Boolean) {
      if (argumentIsNull) {
        // No flag argument is a shortcut for trueish boolean flags
        output.set(key, true);
      } else if (flagArg === 'true') {
        output.set(key, true);
      } else if (flagArg === 'false') {
        output.set(key, false);
      } else {
        throw new ValidationError(
          `Invalid argument for ${longFlag}. '${flagArg}' must be 'true' or 'false'`,
        );
      }
    }
    if (expectedArgType === Type.Number) {
      const asNumber = Utils.tryToNumber(flagArg as string);
      if (asNumber) {
        output.set(key, asNumber);
        continue;
      }
      throw new ValidationError(
        `Invalid type for ${longFlag}. '${
          flagArg as string
        }' is not a valid number`,
      );
    }
    if (expectedArgType === Type.Date) {
      const asDate = Utils.tryToDate(flagArg as string);
      if (asDate) {
        output.set(key, asDate);
        continue;
      }
      throw new ValidationError(
        `Invalid type for ${longFlag}. '${flagArg}' is not a valid date`,
      );
    }
  }
  return Object.fromEntries(output);
};
