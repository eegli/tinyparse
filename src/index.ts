export { CommandBuilder as Parser } from './commands';
export { ValidationError } from './error';

export type { CommandHandler, ErrorHandler, GlobalSetter } from './types';
