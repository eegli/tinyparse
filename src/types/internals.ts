import type { Parser } from '../parser';
import { ErrorParams } from './helpers';

export type LongFlag = `--${string}`;
export type ShortFlag = `-${string}`;

/**
 * The possible default values of a flag option.
 */
export type FlagValue = string | number | boolean | Date;

/**
 * The user settings for a flag option.
 */
export interface FlagOptions<V extends FlagValue, R extends boolean> {
  longFlag: LongFlag;
  shortFlag?: ShortFlag;
  defaultValue: DowncastFlag<V>;
  required?: R;
  description?: string;
  oneOf?: unknown[];
}

export interface FlagOptionsExt<
  V extends FlagValue,
  R extends boolean,
  T = unknown,
> extends FlagOptions<V, R> {
  oneOf: (V | T)[];
}

/**
 * A record of flags and their default values.
 */
export type FlagValueRecord = Record<string, FlagValue>;

/**
 * A map of flags and the settings specified by the user.
 */
export type FlagOptionsMap = Map<string, FlagOptions<FlagValue, boolean>>;

/**
 * Any global settings specified by the user.
 */
export type AnyGlobal = Record<string, unknown>;

/**
 * A pattern that describes the expected arguments of a command.
 */
export type CommandArgPattern = string[] | string | undefined;

/**
 * A map of subcommands and their settings.
 */
export type CommandOptionsMap = Map<
  string,
  Subcommand<FlagValueRecord, AnyGlobal, CommandArgPattern>
>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyHandlerReturn = any | Promise<any>;

type AnyHandler<Options, Globals, Args> = (params: {
  options: Options;
  globals: Globals;
  args: Args;
}) => AnyHandlerReturn;

export type DefaultHandler<
  Options = FlagValueRecord,
  Globals = AnyGlobal,
> = AnyHandler<Options, Globals, string[]>;

export type ErrorHandler = ({ error, usage }: ErrorParams) => void;

/**
 * The settings for a subcommand.
 */
export type Subcommand<Options, Globals, Args> = {
  args: Args;
  description?: string;
  handler: Args extends string[]
    ? // Strict type annotation
      AnyHandler<Options, Globals, DowncastArgs<Args>>
    : // Loose type annotation
      AnyHandler<Options, Globals, string[]>;
};

/**
 * The settings for a subparser.
 */
export type Subparser = {
  description?: string;
  parser: Parser<unknown>;
};

/**
 * A map of subparsers and their settings.
 */
export type SubparserOptionsMap = Map<string, Subparser>;

export type HelpOptions = {
  command?: string;
  longFlag: LongFlag;
  shortFlag?: ShortFlag;
};

export type VersionOptions = {
  version: string;
  command?: string;
  longFlag: LongFlag;
  shortFlag?: ShortFlag;
};
export interface MetaOptions {
  command?: string;
  summary?: string;
  help?: HelpOptions;
  version?: VersionOptions;
}

export type DowncastArgs<T extends string[]> = {
  [P in keyof T]: string;
};

export type DowncastFlag<T extends FlagValue> = T extends string
  ? string
  : T extends number
    ? number
    : T extends boolean
      ? boolean
      : T;
