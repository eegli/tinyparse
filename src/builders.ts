import { transformArgv } from './argv';
import { validateCommandArgs } from './commands';
import { collect } from './flags';
import {
  Subcommand,
  Downcast,
  FlagOptionMap,
  FlagRecord,
  FlagOptions,
  FlagArgValue,
  CommandOptionMap,
  CommandArgPattern,
} from './types';

export class CommandBuilder<F extends FlagRecord = Record<never, never>> {
  #flags: FlagOptionMap = new Map();
  #commands: CommandOptionMap<F> = new Map();

  public subcommand<P extends CommandArgPattern>(
    command: string,
    opts: Subcommand<F, P>,
  ) {
    this.#commands.set(command, opts);
    return this;
  }

  public flag<T extends string, V extends FlagArgValue>(
    flag: T,
    opts: FlagOptions<V>,
  ) {
    this.#flags.set(flag, opts);
    // TODO: Figure out how to make this typecheck properly
    return this as unknown as CommandBuilder<F & Record<T, Downcast<V>>>;
  }

  public build() {
    return new DispatchBuilder<F>(this.#flags, this.#commands);
  }
}

export class DispatchBuilder<F extends FlagRecord> {
  #handler?: (flags: F, positionals: string[]) => void;
  #flags: FlagOptionMap;
  #commands: CommandOptionMap<F>;

  constructor(flags: FlagOptionMap, commands: CommandOptionMap<F>) {
    this.#flags = flags;
    this.#commands = commands;
  }

  public default<T extends F>(
    handler: T extends FlagRecord
      ? (flags: T, positionals: string[]) => void
      : never,
  ) {
    this.#handler = handler;
    return this;
  }

  public parse(argv: string[]): {
    flags: F;
    call: () => void;
  } {
    const [flagMap, positionals] = transformArgv(argv);
    const flags = collect(flagMap, this.#flags) as F;

    const [command, ...args] = positionals;
    const commandOpts = this.#commands.get(command);

    if (command && commandOpts) {
      validateCommandArgs(command, commandOpts, args);
      const call = commandOpts.handler.bind(this, flags, args);
      return {
        flags,
        call,
      };
    }
    if (this.#handler) {
      const call = this.#handler?.bind(this, flags, positionals);
      return {
        flags,
        call,
      };
    }
    return {
      flags,
      call: () => {},
    };
  }
}
