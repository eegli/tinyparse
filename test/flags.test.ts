import { ValidationError } from '../src';
import { collectFlags } from '../src/flags';
import { FlagOptions, FlagOptionsMap, FlagValue } from '../src/types/internals';

describe('flags', () => {
  test('replaces defaults and ignores unknown', () => {
    const options: FlagOptionsMap = new Map([
      ['foo', { longFlag: '--foo', defaultValue: 'default' }],
      ['bar', { longFlag: '--bar', shortFlag: '-b', defaultValue: 'default' }],
      ['baz', { longFlag: '--baz', defaultValue: 'default' }],
    ]);
    const input: Map<string, string | null> = new Map([
      ['--foo', 'bar'],
      ['-b', 'qux'],
    ]);
    const collected = collectFlags(input, options);
    expect(collected).toStrictEqual({ foo: 'bar', bar: 'qux', baz: 'default' });
  });
  test('allows boolean shortcuts and true, false', () => {
    const options: FlagOptionsMap = new Map([
      ['foo', { longFlag: '--foo', defaultValue: false }],
      ['bar', { longFlag: '--bar', defaultValue: false }],
      ['baz', { longFlag: '--baz', defaultValue: true }],
    ]);
    const input: Map<string, string | null> = new Map<string, string | null>([
      ['--foo', null],
      ['--bar', 'true'],
      ['--baz', 'false'],
    ]);
    const collected = collectFlags(input, options);
    expect(collected).toStrictEqual({ foo: true, bar: true, baz: false });
  });
  test('converts valid dates and numbers', () => {
    const options: FlagOptionsMap = new Map([
      ['foo', { longFlag: '--foo', defaultValue: 0 }],
      ['bar', { longFlag: '--bar', defaultValue: new Date() }],
    ]);
    const input: Map<string, string | null> = new Map<string, string | null>([
      ['--foo', '1'],
      ['--bar', '2023'],
    ]);
    const collected = collectFlags(input, options);
    expect(collected).toStrictEqual({
      foo: 1,
      bar: new Date('2023'),
    });
  });
  test('rejects for missing required', () => {
    const options: FlagOptionsMap = new Map([
      ['foo', { longFlag: '--foo', defaultValue: 'default', required: true }],
    ]);
    const input: Map<string, string | null> = new Map([]);
    expect(() => collectFlags(input, options)).toThrow(ValidationError);
    expect(() => collectFlags(input, options)).toThrow(
      'Missing required option --foo',
    );
  });
  test('rejects for invalid types', () => {
    const inputs: [[string, string | null], FlagOptions<FlagValue>][] = [
      [['--foo', null], { longFlag: '--foo', defaultValue: 'string' }],
      [['--foo', 'bar'], { longFlag: '--foo', defaultValue: true }],
      [['--foo', 'bar'], { longFlag: '--foo', defaultValue: 1 }],
      [['--foo', 'bar'], { longFlag: '--foo', defaultValue: new Date() }],
    ];
    for (const [input, option] of inputs) {
      const inputs = new Map([input]);
      const options = new Map([[option.longFlag, option]]);

      expect(() => collectFlags(inputs, options)).toThrow(ValidationError);
      expect(() => collectFlags(inputs, options)).toThrowErrorMatchingSnapshot(
        `<${input.join(' ')}>`,
      );
    }
  });
  test('respects oneof setting', () => {
    const options: FlagOptionsMap = new Map([
      ['foo', { longFlag: '--foo', defaultValue: 'a', oneOf: ['a', 'b'] }],
      ['bar', { longFlag: '--bar', defaultValue: 0, oneOf: [0, 1] }],
    ]);
    const input = new Map([['--foo', 'c']]);
    expect(() => collectFlags(input, options)).toThrow(ValidationError);
    expect(() => collectFlags(input, options)).toThrow(
      'Invalid value "c" for option --foo, expected one of: a, b',
    );
    expect(() => collectFlags(new Map([['--bar', '2']]), options)).toThrow(
      'Invalid value "2" for option --bar, expected one of: 0, 1',
    );
  });
});
