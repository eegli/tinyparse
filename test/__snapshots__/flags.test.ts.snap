// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flags rejects for invalid types: <--foo > 1`] = `"--foo expects an argument"`;

exports[`flags rejects for invalid types: <--foo bar> 1`] = `"Invalid argument for --foo. 'bar' must be 'true' or 'false'"`;

exports[`flags rejects for invalid types: <--foo bar> 2`] = `"Invalid type for --foo. 'bar' is not a valid number"`;

exports[`flags rejects for invalid types: <--foo bar> 3`] = `"Invalid type for --foo. 'bar' is not a valid date"`;
